### ISSUE TITLE
Dependent Pair and Type Function

### ISSUE THREAD

**Issue by xekoukou** [idris-lang/Idris-dev#2981]
On 0.10, this doesn't type check:

```
typeF : String -> Type
typeF "String" = String
typeF "Integer" = Integer

var : Sigma String typeF
var = MkSigma "String" "Hello"
```

**Comment by david-christiansen** [comment]
I wouldn't expect this to type check. There are two reasons:

1. The rules for automatic implicit quantification in type signatures have changed. If you want to refer to something with a lower-case name that isn't being applied to something else, you now need to qualify it with a namespace. Your `var` is thus `{typeF : _} -> Sigma String typeF`, not `Sigma String Main.typeF` as you probably intend.
2. Idris won't reduce functions in types that it doesn't know to be total. Your function `typeF` is manifestly partial, as it only covers two of the potentially infinite number of elements of `String`. To make this work, you'll need a default case at the bottom. It might be worth using a custom datatype with constructors for the cases you care about instead of using strings here.

In other words, Idris seems to be doing what it should here.

In the future, could you please include the complete code and the type errors that you get when you report that something doesn't check that you think should? That makes it easier to work with the bug reports. Thanks!

**Comment by xekoukou** [comment]
Thus, this bug could be transformed into a documentation bug.

A. I checked and you have explained the new changes in automatic implicit quantification to multiple people.  [example.](https://github.com/idris-lang/Idris-dev/issues/2804) Could there be a better way?

Because the language changes and because I am learning the language, when something doesn't work, I use the tutorial and the other documents as a reference. Most of the times I search the documents with error messages or similar concepts I have learned before. The problem with this technique is that even if the documentation has changed, the search method I use might not catch the new change.

So we have two types of people, those that arrive after a language change and those that arrived before. The first will read the documents from the beginning till the end. The others will not. May I propose then that simple examples on the changes be provided outside of the main documents?

Keep in mind that if you provide those examples but do not change the main documents in time, the new people will still have trouble since they will not read those out of band examples. 

B. There does't seem to be a documentation on reduction of functions to types or of this new quantification method.

**Comment by xekoukou** [comment]
This pattern (publishing updates and subscribing to them) seems familiar to me: http://zguide.zeromq.org/page:all#Reliable-Pub-Sub-Clone-Pattern

**Comment by david-christiansen** [comment]
I'm not sure what the relevance of ZeroMQ to Idris name scoping and totality rules is - have you posted a comment on the wrong issue?

With regards to docs, they do state that Idris only reduces known-total operations in the type checker. [Example](http://docs.idris-lang.org/en/latest/faq/faq.html#evaluation-at-the-repl-doesn-t-behave-as-i-expect-what-s-going-on), though this is fairly thoroughly buried.

Language changes are published in the [release notes](http://www.idris-lang.org/idris-0-9-20-released/). The second point describes this change to automatic implicit quantification. I recommend scanning the notes for each release, as that's precisely where we list the user-facing changes.

**Comment by david-christiansen** [comment]
I opened #2983 for the difficulty of finding the information about totality and reduction in the type checker on the docs site.

**Comment by xekoukou** [comment]
The zeromq examples deals with the distributed problem of sending new
updates on a number of clients. One of the problem it has is that some
clients might subscribe at a later time, thus a copy of all previous
updates needs to be kept and send. All clients need to see the same state
eventually. This relates to our problem of updating the language
specification and transmitting that state to all the people interested in
Idris.

Yes, they do say those things but it is very difficult to find them. How do
you find something that you do not know its name? For example the words
"automatic implicit quantification" are alien to me. For the same reason, I
won't be able to understand the changes from the release notes.
